### PR TITLE
User Defined Timestamp APIs

### PR DESCRIPTION
**Pull request by muthukrishnan24**
Adds bindings for [User Defined Timestamp APIs](https://github.com/facebook/rocksdb/wiki/User-defined-Timestamp-(Experimental))

### CODE REVIEW

**Comment by muthukrishnan24** [comment]
@linxGnu

**Comment by linxGnu** [comment]
I am sorry for late reply due to my job. I will take a look and feedback soon.

thank you for the contribution

File: db.go [hunk 1]
```diff
@@ -1307,6 +1547,36 @@ func (db *DB) DeleteFileInRangeCF(cf *ColumnFamilyHandle, r Range) (err error) {
...
+// If another thread updates full_history_ts_low concurrently to a higher
+// timestamp than the requested ts_low, a try again error will be returned.
+func (db *DB) IncreaseFullHistoryTsLow(handle *ColumnFamilyHandle, ts []byte) (err error) {
+	var cErr *C.char
+
+	cTs := byteToChar(ts)
+	C.rocksdb_increase_full_history_ts_low(db.c, handle.c, cTs, C.size_t(len(ts)), &cErr)
+
+	err = fromCError(cErr)
+	return err
```
Comment (linxGnu) [inline db.go:1561]:
```go
return err
```
->
```go
return // no argument
```

File: db.go [hunk 2]
```diff
@@ -1307,6 +1547,36 @@ func (db *DB) DeleteFileInRangeCF(cf *ColumnFamilyHandle, r Range) (err error) {
...
+		cTsLen C.size_t
+	)
+
+	cTs := C.rocksdb_get_full_history_ts_low(db.c, handle.c, &cTsLen, &cErr)
+	if err = fromCError(cErr); err == nil {
+		slice = NewSlice(cTs, cTsLen)
+	}
+
+	err = fromCError(cErr)
+	return slice, err
```
Comment (linxGnu) [inline db.go:1577]: ditto

File: iterator.go [hunk 3]
```diff
@@ -63,6 +64,16 @@ func (iter *Iterator) Key() *Slice {
...
 }
 
+// Timestamp returns the timestamp in the database the iterator currently holds.
+func (iter *Iterator) Timestamp() *Slice {
+	var cLen C.size_t
+	cTs := C.rocksdb_iter_timestamp(iter.c, &cLen)
+	if cTs == nil {
+		return nil
+	}
+	return &Slice{cTs, cLen, true}
```
Comment (linxGnu) [inline iterator.go:74]: How about init with field name:
```go
return &Slice{
   field1: ... 
   field2: ...
```

File: options_compaction.go [hunk 4]
```diff
@@ -33,6 +33,8 @@ const (
 // CompactRangeOptions represent all of the available options for compact range.
 type CompactRangeOptions struct {
 	c *C.rocksdb_compactoptions_t
+
+	fullHistoryTsLow []byte
```
Comment (linxGnu) [inline options_compaction.go:37]: How about removing this field. Seem like we do not need to track this. Its invoker (caller) responsibility

File: transaction.go [hunk 5]
```diff
@@ -467,3 +467,11 @@ func (transaction *Transaction) RebuildFromWriteBatchWI(wb *WriteBatchWI) (err e
 	}
 	return
 }
+
+func (transaction *Transaction) SetCommitTimestamp(ts uint64) {
```
Comment (linxGnu) [inline transaction.go:471]: Please add comment for this

File: transaction.go [hunk 6]
```diff
@@ -467,3 +467,11 @@ func (transaction *Transaction) RebuildFromWriteBatchWI(wb *WriteBatchWI) (err e
...
+	C.rocksdb_transaction_set_commit_timestamp(transaction.c, C.uint64_t(ts))
+}
+
+func (transaction *Transaction) SetReadTimestampForValidation(ts uint64) {
```
Comment (linxGnu) [inline transaction.go:475]: Please add comment for this

**Review by linxGnu — CHANGES_REQUESTED** [review]

**Comment by muthukrishnan24** [comment]
Made the requested changes. Please check @linxGnu

**Review by linxGnu — CHANGES_REQUESTED** [review]

**Comment by linxGnu** [comment]
@muthukrishnan24 

Almost LGTM, just 2 request changes above.

And could you please add more `tests`. It's important to ensure ops with timestamp works. AFAIK, `timestamp` feature is currently under experiment?

**Comment by muthukrishnan24** [comment]
> AFAIK, timestamp feature is currently under experiment?

taken from [here](https://groups.google.com/g/rocksdb/c/1_IWk1lf7GE)
 ```
It's still experimental in the sense that:

    - it's currently supported in a subset of RocksDB features: Get, Put, WriteBatch (with only Put and Delete), Delete, Iterator, basic Garbage collection not necessarily the most efficient implementation), and
    - new APIs and/or options may be added, and
    - (less likely) existing user-defined timestamp APIs and/or options are also subject to change in future RocksDB releases
```

**Comment by linxGnu** [comment]
@muthukrishnan24 thank you for the info.

could you please fix my comments above and especially: adding tests

**Comment by muthukrishnan24** [comment]
Fixed your comments, Added test cases. Please change @linxGnu

**Review by linxGnu — APPROVED** [review]

**Comment by linxGnu** [comment]
Its absolutely great job @muthukrishnan24 
I will merge your PR and prepare for next release.

**Pull request merged**